icu4c/source/i18n/unitsdata.cpp

```diff
@@ -0,0 +1,194 @@
+
+    // TODO(review): felt like a hack: provides easy access to the most recent
+    // baseUnit. This hack is another point making me wonder if doing this
+    // ResourceSink thing is worthwhile. Functional style is not more verbose,
+    // and IMHO more readable than this object-based approach where the output
+    // seems/feels like a side-effect.
+    CharString fLastBaseUnit;
+};
+
+/**
+ * Collects conversion information for a "single unit" (a unit whose complexity
+ * is UMEASURE_UNIT_SINGLE).
+ *
+ * This function currently only supports higher-dimensionality input units if
+ * they map to "single unit" output units. This means it don't support
+ * square-bar, one-per-bar, square-joule or one-per-joule. (Some unit types in
+ * this class: volume, consumption, torque, force, pressure, speed,
+ * acceleration, and more).
+ *
+ * TODO(hugovdm): maybe find and share (in documentation) a solid argument for
+ * why these kinds of input units won't be needed with higher dimensionality? Or
+ * start supporting them... Also: add unit tests demonstrating the
+ * U_ILLEGAL_ARGUMENT_ERROR returned for such units.
+ *
+ * @param unit The input unit. Its complexity must be UMEASURE_UNIT_SINGLE, but
+ * it may have a dimensionality != 1.
+ * @param converUnitsBundle A UResourceBundle instance for the convertUnits
+ * resource.
+ * @param convertSink The ConversionRateDataSink through which
+ * ConversionRateInfo instances are to be collected.
+ * @param status The standard ICU error code output parameter.
+ */
+void processSingleUnit(const MeasureUnit &unit, const UResourceBundle *convertUnitsBundle,
+                       ConversionRateDataSink &convertSink, UErrorCode &status) {
+    if (U_FAILURE(status)) return;
+    int32_t dimensionality = unit.getDimensionality(status);
+
+    // Fetch the relevant entry in convertUnits.
+    MeasureUnit simple = unit;
+    if (dimensionality != 1 || simple.getSIPrefix(status) != UMEASURE_SI_PREFIX_ONE) {
+        simple = unit.withDimensionality(1, status).withSIPrefix(UMEASURE_SI_PREFIX_ONE, status);
+    }
+    ures_getAllItemsWithFallback(convertUnitsBundle, simple.getIdentifier(), convertSink, status);
+}
+
+} // namespace
+
+MaybeStackVector<ConversionRateInfo> U_I18N_API
+getConversionRatesInfo(const MaybeStackVector<MeasureUnit> &units, UErrorCode &status) {
+    MaybeStackVector<ConversionRateInfo> result;
+    if (U_FAILURE(status)) return result;
+
+    LocalUResourceBundlePointer unitsBundle(ures_openDirect(NULL, "units", &status));
+    StackUResourceBundle convertUnitsBundle;
+    ures_getByKey(unitsBundle.getAlias(), "convertUnits", convertUnitsBundle.getAlias(), &status);
+    ConversionRateDataSink convertSink(result);
+
+    for (int i = 0; i < units.length(); i++) {
+        int32_t numSingleUnits;
+        LocalArray<MeasureUnit> singleUnits = units[i]->splitToSingleUnits(numSingleUnits, status);
+
+        for (int i = 0; i < numSingleUnits; i++) {
```

Comment:
The idea with ResourceSink is that you use it once to load a large tree of data.  It's not intended to be used to load subtrees like you're doing here.  For one thing, it's slower, because it appears that you probably have to do a linear search every time (accessing the subtree directly by its key is a faster logarithmic search).  If you're going to be walking the whole tree, then just save everything; don't go looking for specific items many times in a row.

Reply:
Agreed: my code didn't make sense to me. I could iterate over the whole thing with the C-style API too, for simpler code.

I think we concluded that all data should be loaded (it's small enough), if doing so once then and all converters can simply use that data. I'll deal with this change soon (but probably not today).

Reply:
This is related to Hugo pull request: #31

Reply:
Fixed via a new "getAllConversionRates()" method.
I'll take care of getting rid of "getConversionRatesInfo" - but that needs some branch merging or cherry-picking, coordination needed.

Reply:
#31

Reply:
(Resolved in #31)